cou/apps/auxiliary.py

```diff
@@ -204,10 +205,55 @@ class MysqlInnodbCluster(AuxiliaryApplication):
     wait_timeout = LONG_IDLE_TIMEOUT
 
 
-# NOTE (gabrielcocenza): Although CephOSD class is empty now, it will be
-# necessary to add post upgrade plan to set require-osd-release. Registering on
-# AuxiliaryApplication can be easily forgot and ceph-osd can't be instantiated
-# as a normal OpenStackApplication.
 @AppFactory.register_application(["ceph-osd"])
 class CephOSD(AuxiliaryApplication):
     """Application for ceph-osd."""
+
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[COUUnit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[COUUnit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = [
+            PreUpgradeStep(
+                description="Check if colocated nova-compute units had been upgraded.",
+                coro=self._verify_nova_compute(target),
+            )
+        ]
+        steps.extend(super().pre_upgrade_steps(target, units))
+        return steps
+
+    async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
+        """Check if a nova-compute application has upgraded its workload version.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises ApplicationError: When any nova-compute app workload version isn't reached.
+        """
+        units_not_upgraded = []
+        apps = await self.model.get_applications()
+
+        for app in apps:
+            if app.charm != "nova-compute":
+                logger.debug("skipping application %s", app.name)
+                continue
```

Comment:
I'm using if and continue, to avoid having to many nested functionality. So I want to avoid having
```python
for ...
    if ...
        for ...
```
but rather
```python
for ...
    if ...
        continue

    for ...
```

Reply:
I think it's easier to understand the code flow without this continue, but it's just personal taste